File: Projects/PythonProject/Object_Class_Methods/HashCodeExample.py
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    # Overriding __eq__ for logical equality
    def __eq__(self, other):
        if not isinstance(other, Person):
            return False
        return self.name == other.name and self.age == other.age

    # Overriding __hash__ to maintain the contract with __eq__
    def __hash__(self):
        return hash((self.name, self.age))

    # Optional: for better print output
    def __repr__(self):
        return f"Person(name='{self.name}', age={self.age})"

# Main logic to demonstrate hash and equality
def main():
    person1 = Person("John Doe", 30)
    person2 = Person("John Doe", 30)

    print("Hash code of person1:", hash(person1))
    print("Hash code of person2:", hash(person2))
    print("Are person1 and person2 equal?", person1 == person2)
    print("person1:", person1)
    print("person2:", person2)

if __name__ == "__main__":
    main()

File: Projects/JavaProject/src/com/Object_Class_Methods/HashCodeExample.java
package com.Object_Class_Methods;

// Class representing a custom data type
class Person {
    private String name;
    private int age;

    // Constructor to initialize the Person object
    public Person(String name, int age) {
        this.name = name;
        this.age = age;
    }

    // Overriding the hashCode() method
    @Override
    public int hashCode() {
        // Compute hash code based on the object's data
        int result = name.hashCode(); // Hash code of the name field
        result = 31 * result + age;   // Combine with the age field using a prime number
        return result;
    }

    // Overriding equals() method for proper comparison (to maintain contract between equals() and hashCode())
    // @Override
    // public boolean equals(Object obj) {
    //     if (this == obj) {
    //         return true;
    //     }
    //     if (obj == null || getClass() != obj.getClass()) {
    //         return false;
    //     }
    //     Person otherPerson = (Person) obj;
    //     return this.name.equals(otherPerson.name) && this.age == otherPerson.age;
    // }
}

// Main class to demonstrate the hashCode() method
public class HashCodeExample {
    public static void main(String[] args) {
        // Creating two Person objects with the same data
        Person person1 = new Person("John Doe", 30);
        Person person2 = new Person("John Doe", 30);

        // Default and overridden hashCode() behavior
        System.out.println("Hash code of person1: " + person1.hashCode());
        System.out.println("Hash code of person2: " + person2.hashCode());
        System.out.println("Are person1 and person2 equal? " + person1.equals(person2));
        System.out.println(person1);
        System.out.println(person2);
    }
}
